debian/easy-apache.1
.TH "Easy Apache" 1
.SH NAME
easy-apache \- script to easily setup apache server and install ssl certificate(s)
.SH SYNOPSIS
.B easy-apache
[OPTIONS]
.SH DESCRIPTION
.PP
Easy Apache is a simple script that helps you setup an Apache server, configure it completely, and it also helps you in installing SSL certificates for all your Apache sites. It'll ask for a few inputs, like domain name, do you want to install SSL, etc that's it.
.PP
It takes care of configuring Apache, setting up virtual hosts (vhosts), managing different directories for each site, enabling site, disabling sites, and even installing an SSL certificate to your Apache sites.
.SH OPTIONS
.TP
.BR \-s ", " \-\-ssl
Setup domain with SSL certificates
.TP
.BR \-h ", " \-\-help
Help (shows available commands)
.TP
.BR \-v ", " \-\-version
Check easy-apache version
.SH EXAMPLES
.TP
.BR easy-apache " " \-s example.com 
Setup domain with ssl certificates
.SH AUTHORS
.PP
(C) Copyright 2020 R Pavan Kumar <realpvn@gmail.com>
.PP
This is free software; see the source for copying conditions. There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH BUGS
Report bugs on the github repo for easy-apache (https://github.com/realpvn/easy-apache.git). Open a issue and I will reply/fix as soon as I can.
.SH CONTRIBUTE
.PP
You can contribute to Easy Apache here - https://github.com/realpvn/easy-apache.git. If your contributions are good/big your name might be added to AUTHORS section

debian/changelog
easy-apache (3.1-1) focal; urgency=medium

  * Fixed small issue where banner would not print

 -- Pavan Kumar <realpvn@gmail.com>  Sun, 07 Mar 2021 17:37:44 +0530

easy-apache (3.0-1) focal; urgency=medium

  * Major improvements to how the script runs and handles errors
  * Easier way to add new sites
  * Option to add SSL certificates to already existing sites
  * When domain setup fails, everything done will be reverted, helping keep 
    system clean
  * Fixed version showing not showing as expected
  * Script should be faster than older version

 -- Pavan Kumar <realpvn@gmail.com>  Sun, 07 Mar 2021 16:59:07 +0530

easy-apache (2.13-1) focal; urgency=medium

  * Added man page for more details about easy-apache
  * Improvements while showing banner
  * Removed code which tried to install packages again 

 -- Pavan Kumar <realpvn@gmail.com>  Fri, 16 Oct 2020 21:35:20 +0530

easy-apache (2.12-1) focal; urgency=medium

  * Remove dpkg usage since it will create problems for snap packages 

 -- Pavan Kumar <realpvn@gmail.com>  Wed, 14 Oct 2020 17:46:17 +0530

easy-apache (2.11-1) focal; urgency=medium

  * Remove unneccessary files from user bin

 -- Pavan Kumar <realpvn@gmail.com>  Sat, 26 Sep 2020 18:20:19 +0530

easy-apache (2.10-1) focal; urgency=medium

  * Multiple changes to banner and how it is fetched 

 -- Pavan Kumar <realpvn@gmail.com>  Sat, 26 Sep 2020 14:28:55 +0530

easy-apache (2.9-1) focal; urgency=medium

  * Remove makefile for now as its not neccessary 

 -- Pavan Kumar <realpvn@gmail.com>  Sat, 26 Sep 2020 13:35:48 +0530

easy-apache (2.8-1) focal; urgency=medium

  * Fixed few makefile errors

 -- Pavan Kumar <realpvn@gmail.com>  Sat, 26 Sep 2020 13:19:28 +0530

easy-apache (2.7-1) focal; urgency=medium

  * Major improvements 

 -- Pavan Kumar <realpvn@gmail.com>  Sat, 26 Sep 2020 12:58:41 +0530

easy-apache (2.6-1) focal; urgency=medium

  * Fixed a build error 

 -- Pavan Kumar <realpvn@gmail.com>  Wed, 09 Sep 2020 10:22:52 +0530

easy-apache (2.5-1) focal; urgency=medium

  * Improvements to the way script runs
  * Added better message when SSL installation fails
  * Added a manpage

 -- R Pavan Kumar <realpvn@gmail.com>  Wed, 09 Sep 2020 10:08:42 +0530

easy-apache (2.4-1) focal; urgency=medium

  * Changes to Documentation

 -- R Pavan Kumar <realpvn@gmail.com>  Tue, 08 Sep 2020 08:39:10 +0530

easy-apache (2.3-1) focal; urgency=medium

  * Use certbot from apt repository instead of ppa

 -- R Pavan Kumar <realpvn@gmail.com>  Tue, 08 Sep 2020 07:39:54 +0530

easy-apache (2.2-1) focal; urgency=medium

  * Remove unwanted files

 -- R Pavan Kumar <realpvn@gmail.com>  Mon, 07 Sep 2020 15:30:40 +0530

easy-apache (2.1-1) focal; urgency=medium

  * Fixed server & ssl script missing
  * Combined multiple scripts into one, so execution is easier & less files to worry about

 -- R Pavan Kumar <realpvn@gmail.com>  Mon, 07 Sep 2020 15:03:29 +0530

easy-apache (2.0-1) focal; urgency=medium

  * Initial release

 -- R Pavan Kumar <realpvn@gmail.com>  Sun, 06 Sep 2020 14:03:29 +0530


easy-apache.sh
#!/bin/bash

serverIP=`curl -s icanhazip.com`
banner() {
	curl https://raw.githubusercontent.com/realpvn/easy-apache/master/easy-banner.txt
}

terminalColors () {
	# Colors from - https://gist.github.com/5682077.git
	TC='\e['

	Bold="${TC}1m"    # Bold text only, keep colors
	Undr="${TC}4m"    # Underline text only, keep colors
	Rst="${TC}0m"     # Reset all coloring and style

	# Basic colors
	Black="${TC}30m"
	Red="${TC}31m"
	Green="${TC}32m"
	Yellow="${TC}33m"
	Blue="${TC}34m"
	Purple="${TC}35m"
	Cyan="${TC}36m"
	White="${TC}37m"

	Crossed="\u2718"

	#copy pasted tick symbold because unicode was acting weird
	Ticked="✓"
	Info="!"

	Oper=${Bold}'[ * ]'
	OperSuccess=${Bold}${Green}'[ '${Ticked}' ]'
	OperFailed=${Bold}${Red}'[ '${Crossed}' ]'
	Info=${Bold}${Yellow}'[ '${Info}' ]'
}

printNormal () {
	echo -e "${Oper} $1$Rst";
}

printSuccess() {
	echo -e "${OperSuccess} $1$Rst"
}

printFailed() {
	echo -e "${OperFailed} $1$Rst"
}

printInfo() {
	echo -e "${Info} $1$Rst"
}

help () {
	banner
	echo -e "${Bold}Usage:${Rst}\neasy-apache [options]\n\n${Bold}Options:${Rst}\n-f | --full:\t\tFull setup, default option if none is provided\n-a | --apache:\t\tAdding new site (includes apache install)\n-s | --SSL:\t\tInstall SSL certificate for sites available\n-h | --help:\t\tHelp (shows available commands)\n-v | --version:\t\tCheck easy-apache version${Rst}"
	echo -e "\n${Bold}Example\n${Rst}easy-apache -f\t\t\t#for full installation i.e Apache & SSL certificate\neasy-apache --full\t\t#for full installation i.e Apache & SSL certificate\neasy-apache -a\t\t\t#for installating Apache server\neasy-apache -s\t\t\t#for installating SSL certificate"
}

version() {
	version=`apt-cache show easy-apache | grep Version | cut -d ' ' -f2`
	echo $version
}

apacheInstall () {
	banner
	printInfo "Server Public IP: ${Yellow}${Bold}"${serverIP}${Rst}

	# Updating Server
	printNormal "Updating Server"
	sudo apt update -y &> /dev/null
	printNormal "Almost done, this might take few more mins"
	sudo apt upgrade -y &> /dev/null
	printSuccess "Updated"

	# Cleaning after server update
	sudo apt autoremove -y &> /dev/null
	sudo apt autoclean -y &> /dev/null
	printSuccess "Cleaned after server update"

	allSitesURL=""
	allSitesCount=-1

	# Installing apache2
	printSuccess "Apache Installed"
	printInfo "${Yellow}${Bold}Apache `apache2 -v`${Rst}"

	while true
	do
		printInfo "Setting up new site"
		read -p "Site URL (99 - exit): " siteURL

		if [[ -z "$siteURL" ]]
		then
			printFailed "Site URL empty"
			break
		fi

		if [ $siteURL == 99 ]
		then
			if [ $allSitesCount == -1 ]
			then
				echo "Good byee..."
				exit
			fi
			break
		fi

		re="(https:\/\/)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)"
		if [[ ! $siteURL =~ $re ]]
		then
			printFailed "Invalid URL"
			continue
		fi

		printSuccess "URL Valid"

		if [ -e /etc/apache2/sites-available/$siteURL.conf ]
		then
			printFailed "$siteURL already exists"
			printInfo "Do you want to overwrite $siteURL? (Yy/Nn/99 to exit setup)"
			read overwriteSite
			case $overwriteSite in
				[Yy]* ) #allSitesURL used for printing at last
						allSitesCount=`expr $allSitesCount + 1`
						allSitesURL[$allSitesCount]=$siteURL
						addSite $siteURL
						break;;
				[Nn]* ) continue;;
				[99]* ) break;;
					* ) printInfo "Invalid input. Skipping $siteURL setup"
						continue;;
			esac
		else
			#allSitesURL used for printing at last
			allSitesCount=`expr $allSitesCount + 1`
			allSitesURL[$allSitesCount]=$siteURL
			addSite $siteURL
			break
		fi
	done

	if [ $allSitesCount == -1 ]
	then
		exit
	fi

	if sudo ufw status | grep -q inactive$
	then
		printInfo "UFW is disabled. You need to enable it to continue..."
		
		while true
		do
			read -p "Do you want to enable now (Yy/Nn)? " ufwEnable
			case $ufwEnable in
				[Yy]* ) echo "y" | sudo ufw enable &> /dev/null;
						printSuccess "UFW enabled"
						sudo ufw allow ssh &> /dev/null;
						sudo ufw allow Apache &> /dev/null;
						printSuccess "Allowed SSH & Apache in ufw"
						break;;
				[Nn]* ) printInfo "You cannot view the site until you enable ufw and allow SSH & Apache"; break;;
					* ) echo -e "Please answer yes(Yy) or no(Nn) ";;
			esac
		done
	else
		printSuccess "UFW already enabled"
	fi

	sudo a2dissite 000-default.conf &> /dev/null
	printSuccess "Disabled default site"

	sudo systemctl restart apache2
	printSuccess "Apache restart success"

	#TODO(pavank): try to check if site is available at server ${IP}

	printSuccess "Success! Your site(s) have been added successfully"
	printSuccess "Point your domains A record to $IP and after DNS propagation everything should be working fine.${Rst}"
	echo -e "Sites added and configured are:"

	temp=-1
	while [ $temp != $allSitesCount ]
	do
		if [ -e /etc/apache2/sites-available/${allSitesURL[$temp]}-le-ssl.conf ]
		then
			temp=`expr $temp + 1`
			echo -e "https://${allSitesURL[$temp]}"
			continue
		fi
		temp=`expr $temp + 1`
		echo -e "http://"${allSitesURL[$temp]}
	done
}

addSite () {
	siteURL=$1

	# Used for directory name (which is without domain TLD, example.com site folder would be "example" not "example.com")
	siteNameNoTLD=`echo -e $siteURL | cut -d'.' -f1`

	printInfo "Creating directory & adding permissions"
	sudo mkdir -p /var/www/$siteNameNoTLD
	sudo chown -R $USER:$USER /var/www/$siteNameNoTLD
	sudo chmod -R 755 /var/www
	printSuccess "Done"

	# Create temporary index.html page for viewing
	sudo echo -e "<h1>Server setup by <a href='https://github.com/realpvn/easy-apache.git'>easy-apache</a> (https://github.com/realpvn/easy-apache.git) </h1>" &> /var/www/$siteNameNoTLD/index.html

	printSuccess "Site $siteURL created"
	printNormal "Configuring"
	read -p "Email (leave blank if not required):" siteEmail
	if [ -z $siteEmail ]; then
		siteEmail=dev@localhost
	fi
	echo -e "<VirtualHost *:80>\n\tServerAdmin $siteEmail\n\tServerName $siteURL\n\tServerAlias www.$siteURL\n\tDocumentRoot /var/www/$siteNameNoTLD\n\tErrorLog \${APACHE_LOG_DIR}/error.log\n\tCustomLog \${APACHE_LOG_DIR}/access.log combined\n</VirtualHost>" | sudo tee /etc/apache2/sites-available/$siteURL.conf
	printSuccess "Site $siteURL configured successfully"

	sudo a2ensite $siteURL.conf &> /dev/null
	printSuccess "Enabled configuration for $siteURL"
}

sslInstall () {
	banner
	printNormal "SSL Installation"
	allSitesURL=""
	allSitesCount=-1

	# Install certbot to install and manage SSL
	sudo apt install certbot python3-certbot-apache -y &> /dev/null
	printSuccess "Installed Certbot"

	while true
	do
		read -p "Site URL to add SSL (99 - Exit): " siteName
		if [[ -z "$siteName" ]]
		then
			printFailed "Site URL empty"
			break
		fi

		if [ $siteName == 99 ]
		then
			if [ $allSitesCount == -1 ]
			then
				echo -e "Good byee..."
				exit
			fi
			break
		fi

		re="(https:\/\/)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)"
		if [[ ! $siteName =~ $re ]]
		then
			printFailed "Invalid URL"
			printInfo "Example site URL - example.com"
			continue
		fi

		if [ ! -e /etc/apache2/sites-available/${siteName}.conf ]
		then
			printFailed "${siteName} does not exist"
			printInfo "Use 'easy-apache -a' to add a site"
			exit
		fi

		if [ -e /etc/apache2/sites-available/${siteName}-le-ssl.conf ]
		then
			printFailed "${siteName} already has SSL installed"
			continue
		fi

		allSitesCount=`expr $allSitesCount + 1`
		allSitesURL[$allSitesCount]=$siteName
		
		sudo certbot --apache -d www.$siteName -d $siteName
		if [ -e /etc/apache2/sites-available/$siteName-le-ssl.conf ]
		then
			printSuccess "SSL Successful for $siteName"
			continue
		fi
		printFailed "SSL unsuccessful for $siteName"
		printInfo "Make sure your domains A record is pointing to your IP: $serverIP${Rst}"
		exit
	done

	if [ $allSitesCount == -1 ]
	then
		printFailed "SSL Installation failed"
		printInfo "Reasons for failure"
		printInfo "1. You do not have any sites added (use 'easy-apache -a' to add site)"
		printInfo "2. You already have ssl certificates installed"
		printInfo "Please check & run SSL Installation again"
		exit
	fi

	sudo ufw delete allow 'Apache' &> /dev/null
	sudo ufw allow 'Apache Full' &> /dev/null
	printSuccess "Apache Full allowed in ufw"

	echo -e "SSL added to sites:"
	temp=-1
	while [ $temp != $allSitesCount ]
	do
		if [ -e /etc/apache2/sites-available/${allSitesURL[$temp]}-le-ssl.conf ]
		then
			temp=`expr $temp + 1`
			echo -e "https://${allSitesURL[$temp]}"
			continue
		fi
		temp=`expr $temp + 1`
		echo -e "http://"${allSitesURL[$temp]}
	done
}

terminalColors
if [ "$1" != "" ]
then
    PARAM=$1
    case $PARAM in
        -f | --full 	 ) apacheInstall; sslInstall; exit;;
        -a | --apache 	 ) apacheInstall; exit;;
        -s | --SSL       ) sslInstall; exit;;
        -h | --help      ) help; exit;;
		-v | --version   ) version; exit;;
                        *) help; exit;;
    esac
fi
help